packages/rs-drive-light-client/src/proof/from_proof.rs

```diff
@@ -0,0 +1,208 @@
+use std::fmt::Debug;
+
+use crate::Error;
+use dapi_grpc::platform::v0::{self as platform};
+use dpp::prelude::{Identifier, Identity};
+pub use drive::drive::verify::RootHash;
+use drive::drive::Drive;
+
+use super::verify::verify_tenderdash_proof;
+
+// #[cfg(feature = "mockall")]
+
+/// Create an object based on proof received from DAPI
+pub trait FromProof<Req, Resp> {
+    /// Parse and verify the received proof and retrieve the requested object, if any.
+    ///
+    /// # Arguments
+    ///
+    /// * `request`: The request sent to the server.
+    /// * `response`: The response received from the server.
+    /// * `provider`: A callback implementing [QuorumInfoProvider] that provides quorum details required to verify the proof.
+    ///
+    /// # Returns
+    ///
+    /// * `Ok(Some(object))` when the requested object was found in the proof.
+    /// * `Ok(None)` when the requested object was not found in the proof; this can be interpreted as proof of non-existence.
+    /// * `Err(Error)` when either the provided data is invalid or proof validation failed.
+    fn maybe_from_proof(
+        request: &Req,
+        response: &Resp,
+        provider: Box<dyn QuorumInfoProvider>,
+    ) -> Result<Option<Self>, Error>
+    where
+        Self: Sized;
+
+    /// Retrieve the requested object from the proof.
+    ///
+    /// Runs full verification of the proof and retrieves enclosed objects.
+    ///
+    /// This method uses [`maybe_from_proof()`] internally and throws an error if the requested object does not exist in the proof.
+    ///
+    /// # Arguments
+    ///
+    /// * `request`: The request sent to the server.
+    /// * `response`: The response received from the server.
+    /// * `provider`: A callback implementing [QuorumInfoProvider] that provides quorum details required to verify the proof.
+    ///
+    /// # Returns
+    ///
+    /// * `Ok(object)` when the requested object was found in the proof.
+    /// * `Err(Error::DocumentMissingInProof)` when the requested object was not found in the proof.
+    /// * `Err(Error)` when either the provided data is invalid or proof validation failed.
+    fn from_proof(
+        request: &Req,
+        response: &Resp,
+        provider: Box<dyn QuorumInfoProvider>,
+    ) -> Result<Self, Error>
+    where
+        Self: Sized,
+    {
+        Self::maybe_from_proof(request, response, provider)?.ok_or(Error::DocumentMissingInProof)
+    }
+}
+
+/// `QuorumInfoProvider` trait provides an interface to fetch quorum related information, required to verify the proof.
+///
+/// Developers should implement this trait to provide required quorum details to [FromProof] implementations.
+///
+/// It defines a single method `get_quorum_public_key` which retrieves the public key of a given quorum.
+#[uniffi::export(callback_interface)]
+#[cfg_attr(feature = "mock", mockall::automock)]
+pub trait QuorumInfoProvider: Send + Sync {
```

Comment:
We have trait and callback interface just to get a quorum public key that might be just passed as a parameter, how sure we are that this trait will grow? If so, do we really need to invert how dependencies are taken instead of having each method to have their own requirements listed as arguments?

I just don't get this whole dynamic story with request object and vague provider instead of statically known calls with their own dependencies, do we need this runtime flexibility?